swarm/src/lib.rs

```diff
@@ -1018,88 +1018,92 @@ where
                 Poll::Ready(pool_event) => {
                     if let Some(swarm_event) = this.handle_pool_event(pool_event) {
                         return Poll::Ready(swarm_event);
                     }
-                }
-            };
 
-            // After the network had a chance to make progress, try to deliver
-            // the pending event emitted by the behaviour in the previous iteration
-            // to the connection handler(s). The pending event must be delivered
-            // before polling the behaviour again. If the targeted peer
-            // meanwhie disconnected, the event is discarded.
-            if let Some((peer_id, handler, event)) = this.pending_event.take() {
-                match handler {
-                    PendingNotifyHandler::One(conn_id) => {
-                        if let Some(mut conn) = this.pool.get_established(conn_id) {
-                            if let Some(event) = notify_one(&mut conn, event, cx) {
-                                this.pending_event = Some((peer_id, handler, event));
-                                if listeners_not_ready && connections_not_ready {
-                                    return Poll::Pending;
-                                } else {
-                                    continue;
-                                }
-                            }
-                        }
-                    }
-                    PendingNotifyHandler::Any(ids) => {
-                        if let Some((event, ids)) =
-                            notify_any::<_, _, TBehaviour>(ids, &mut this.pool, event, cx)
-                        {
-                            let handler = PendingNotifyHandler::Any(ids);
-                            this.pending_event = Some((peer_id, handler, event));
-                            if listeners_not_ready && connections_not_ready {
-                                return Poll::Pending;
-                            } else {
-                                continue;
-                            }
-                        }
-                    }
+                    continue;
                 }
-            }
-
-            debug_assert!(this.pending_event.is_none());
-
-            let behaviour_poll = {
-                let mut parameters = SwarmPollParameters {
-                    local_peer_id: &this.local_peer_id,
-                    supported_protocols: &this.supported_protocols,
-                    listened_addrs: &this.listened_addrs,
-                    external_addrs: &this.external_addrs,
-                };
-                this.behaviour.poll(cx, &mut parameters)
             };
 
-            match behaviour_poll {
-                Poll::Pending if listeners_not_ready && connections_not_ready => {
-                    return Poll::Pending
-                }
+            // Poll the listener(s) for new connections.
+            match ListenersStream::poll(Pin::new(&mut this.listeners), cx) {
```

Comment:
If `ListenersStream` would actually implement `Stream` then we could utilise `poll_unpin` here.

Reply:
True, though we would have to handle the `None` case which can never happen. https://github.com/rust-lang/futures-rs/issues/1857 would be very useful here.